scene/animation/animation_blend_tree.cpp

```diff
@@ -86,46 +86,55 @@ double AnimationNodeAnimation::_process(double p_time, bool p_seek, bool p_is_ex
 	Ref<Animation> anim = ap->get_animation(animation);
 	double anim_size = (double)anim->get_length();
 	double step = 0.0;
-	double prev_time = cur_time;
 	Animation::LoopedFlag looped_flag = Animation::LOOPED_FLAG_NONE;
 	bool node_backward = play_mode == PLAY_MODE_BACKWARD;
 
 	if (p_seek) {
-		step = p_time - cur_time;
-		cur_time = p_time;
+		step = p_time - cumulative_time;
 	} else {
 		p_time *= backward ? -1.0 : 1.0;
-		cur_time = cur_time + p_time;
 		step = p_time;
 	}
 
+	// cur_time is the modulated version of the overall time.
+	double cur_time = 0.0;
 	bool is_looping = false;
 	if (anim->get_loop_mode() == Animation::LOOP_PINGPONG) {
 		if (!Math::is_zero_approx(anim_size)) {
-			if (prev_time >= 0 && cur_time < 0) {
+			double mod_prev_time = Math::fposmod(cumulative_time, anim_size);
+			cur_time = mod_prev_time + step;
+
+			if (mod_prev_time >= 0 && cur_time < 0) {
 				backward = !backward;
 				looped_flag = node_backward ? Animation::LOOPED_FLAG_END : Animation::LOOPED_FLAG_START;
 			}
-			if (prev_time <= anim_size && cur_time > anim_size) {
+			if (mod_prev_time <= anim_size && cur_time > anim_size) {
 				backward = !backward;
 				looped_flag = node_backward ? Animation::LOOPED_FLAG_START : Animation::LOOPED_FLAG_END;
 			}
-			cur_time = Math::pingpong(cur_time, anim_size);
+			cumulative_time = cur_time = Math::pingpong(cur_time, anim_size);
```

Comment:
```suggestion
			cur_time = Math::pingpong(cur_time, anim_size);
			cumulative_time = cur_time;
```
This chained assignment hurts readability and is very rarely done in the current source

Reply:
Removed the chained assignment now.